File: event_countdown/lib/features/pomdoro/services/pomodoro_service.dart
import 'dart:convert';

import 'package:amplify_flutter/amplify_flutter.dart';
import 'package:event_countdown/core/service/auth_service.dart';
import 'package:event_countdown/features/models/pomdoro_model.dart';
import 'package:event_countdown/features/pomdoro/services/pomodoro_repository.dart';

/// Service class for handling all Pomodoro-related API operations.
/// Backend uses a single-table design; all requests require the current user's id (Cognito sub).
class PomodoroService implements PomodoroRepository {
  static const String _apiName = 'CountdownApi';
  static const String _pomodorosEndpoint = '/pomodoros';

  final AuthService authService = AuthService();

  /// Fetches all pomodoro sessions for a todo.
  /// Returns empty list on auth/API errors.
  @override
  Future<List<Pomodoro>> getPomodoros(String todoId) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) {
        safePrint('User is not signed in');
        throw Exception('Not signed in');
      }

      final response = await Amplify.API
          .get(
            _pomodorosEndpoint,
            apiName: _apiName,
            queryParameters: {'userId': userId, 'todoId': todoId},
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      if (response.statusCode != 200) {
        safePrint('Get pomodoros failed: ${response.statusCode}');
        throw Exception('Failed to load pomodoros: ${response.statusCode}');
      }

      final body = response.decodeBody();
      final list = jsonDecode(body) as List<dynamic>;
      return list
          .map((e) => Pomodoro.fromJson(e as Map<String, dynamic>))
          .toList();
    } on ApiException catch (e) {
      safePrint('Pomodoro get API error: $e');
      rethrow;
    } catch (e) {
      safePrint('Error fetching pomodoros: $e');
      rethrow;
    }
  }

  /// Adds a new pomodoro session for a todo.
  /// Returns the new pomodoro id if successful, null otherwise.
  /// Backend expects [userId], [todoId], [title], [timerDurationInMinutes].
  @override
  Future<String?> addPomodoro({
    required String todoId,
    required String title,
    required int timerDurationInMinutes,
  }) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) {
        safePrint('User is not signed in');
        return null;
      }

      final response = await Amplify.API
          .post(
            _pomodorosEndpoint,
            apiName: _apiName,
            body: HttpPayload.json({
              'userId': userId,
              'todoId': todoId,
              'title': title,
              'timerDurationInMinutes': timerDurationInMinutes,
            }),
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      safePrint(
        'Pomodoro added (${response.statusCode}): ${response.decodeBody()}',
      );

      if (response.statusCode != 200 && response.statusCode != 201) {
        return null;
      }

      final body = response.decodeBody();
      final json = jsonDecode(body) as Map<String, dynamic>;
      return json['id'] as String?;
    } on ApiException catch (e) {
      safePrint('Pomodoro create API error: $e');
      return null;
    } catch (e) {
      safePrint('Error creating pomodoro: $e');
      return null;
    }
  }

  /// Starts the timer for a pomodoro by saving a startedAt timestamp in the DB.
  /// Returns the server-assigned [startedAt] DateTime on success.
  @override
  Future<DateTime?> startPomodoro({
    required String pomodoroId,
    required String todoId,
  }) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) return null;

      final response = await Amplify.API
          .put(
            _pomodorosEndpoint,
            apiName: _apiName,
            body: HttpPayload.json({
              'userId': userId,
              'todoId': todoId,
              'pomodoroId': pomodoroId,
              'action': 'start',
            }),
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      if (response.statusCode != 200) {
        safePrint('Start pomodoro failed: ${response.statusCode}');
        return null;
      }

      final body = jsonDecode(response.decodeBody()) as Map<String, dynamic>;
      return DateTime.parse(body['startedAt'] as String);
    } catch (e) {
      safePrint('Error starting pomodoro: $e');
      return null;
    }
  }

  /// Pauses a running pomodoro. The server accumulates elapsed time and clears
  /// startedAt. Returns the new total [elapsedSeconds], or null on failure.
  @override
  Future<int?> pausePomodoro({
    required String pomodoroId,
    required String todoId,
  }) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) return null;

      final response = await Amplify.API
          .put(
            _pomodorosEndpoint,
            apiName: _apiName,
            body: HttpPayload.json({
              'userId': userId,
              'todoId': todoId,
              'pomodoroId': pomodoroId,
              'action': 'pause',
            }),
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      if (response.statusCode != 200) return null;

      final body = jsonDecode(response.decodeBody()) as Map<String, dynamic>;
      return (body['elapsedSeconds'] as num).toInt();
    } catch (e) {
      safePrint('Error pausing pomodoro: $e');
      return null;
    }
  }

  /// Resets a pomodoro back to its initial state: clears startedAt,
  /// sets elapsedSeconds to 0, and sets status to "stopped".
  @override
  Future<bool> resetPomodoro({
    required String pomodoroId,
    required String todoId,
  }) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) return false;

      final response = await Amplify.API
          .put(
            _pomodorosEndpoint,
            apiName: _apiName,
            body: HttpPayload.json({
              'userId': userId,
              'todoId': todoId,
              'pomodoroId': pomodoroId,
              'action': 'reset',
            }),
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      return response.statusCode == 200;
    } catch (e) {
      safePrint('Error resetting pomodoro: $e');
      return false;
    }
  }

  /// Deletes a pomodoro session.
  /// Returns true if successful, false otherwise.
  @override
  Future<bool> deletePomodoro({
    required String pomodoroId,
    required String todoId,
  }) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) return false;

      final response = await Amplify.API
          .delete(
            _pomodorosEndpoint,
            apiName: _apiName,
            queryParameters: {
              'userId': userId,
              'todoId': todoId,
              'pomodoroId': pomodoroId,
            },
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      if (response.statusCode != 200) {
        safePrint('Delete pomodoro failed: ${response.statusCode}');
        return false;
      }
      return true;
    } on ApiException catch (e) {
      safePrint('Delete pomodoro API error: $e');
      return false;
    } catch (e) {
      safePrint('Error deleting pomodoro: $e');
      return false;
    }
  }

  /// Marks a pomodoro as completed in the DB.
  @override
  Future<bool> completePomodoro({
    required String pomodoroId,
    required String todoId,
  }) async {
    try {
      final token = await authService.getIdToken();
      final userId = await authService.getUserId();
      if (token == null || userId == null) return false;

      final response = await Amplify.API
          .put(
            _pomodorosEndpoint,
            apiName: _apiName,
            body: HttpPayload.json({
              'userId': userId,
              'todoId': todoId,
              'pomodoroId': pomodoroId,
              'action': 'complete',
            }),
            headers: {'Authorization': 'Bearer $token'},
          )
          .response;

      return response.statusCode == 200;
    } catch (e) {
      safePrint('Error completing pomodoro: $e');
      return false;
    }
  }
}

File: event_countdown_web2/src/app/services/PomodoroService.ts
"use client";

import { del, get, post, put } from "aws-amplify/api";
import { bearerAuthHeaders, getUserId } from "@/src/app/services/AuthService";
import type { PomodoroRecord } from "@/src/features/pomodoro/pomodoro-model";

export async function fetchPomodorosForTodo(
  todoId: string,
): Promise<PomodoroRecord[]> {
  const userId = await getUserId();
  if (!userId) throw new Error("Not signed in");

  const headers = await bearerAuthHeaders();
  const { body, statusCode } = await get({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      queryParams: { userId, todoId },
    },
  }).response;

  if (statusCode !== 200) {
    throw new Error(`Failed to load pomodoros (${String(statusCode)})`);
  }

  const json = (await body.json()) as unknown;
  if (!Array.isArray(json)) return [];
  return json as PomodoroRecord[];
}

export async function createPomodoro(input: {
  todoId: string;
  title: string;
  timerDurationInMinutes: number;
}): Promise<string | null> {
  const userId = await getUserId();
  if (!userId) return null;

  const headers = await bearerAuthHeaders();
  const { body, statusCode } = await post({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      body: {
        userId,
        todoId: input.todoId,
        title: input.title,
        timerDurationInMinutes: input.timerDurationInMinutes,
      },
    },
  }).response;

  if (statusCode !== 200 && statusCode !== 201) return null;
  const json = (await body.json()) as { id?: string };
  return json.id ?? null;
}

export async function startPomodoro(input: {
  pomodoroId: string;
  todoId: string;
}): Promise<string | null> {
  const userId = await getUserId();
  if (!userId) return null;

  const headers = await bearerAuthHeaders();
  const { body, statusCode } = await put({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      body: {
        userId,
        todoId: input.todoId,
        pomodoroId: input.pomodoroId,
        action: "start",
      },
    },
  }).response;

  if (statusCode === 409) return null;
  if (statusCode !== 200) return null;
  const json = (await body.json()) as { startedAt?: string };
  return json.startedAt ?? null;
}

export async function pausePomodoro(input: {
  pomodoroId: string;
  todoId: string;
}): Promise<number | null> {
  const userId = await getUserId();
  if (!userId) return null;

  const headers = await bearerAuthHeaders();
  const { body, statusCode } = await put({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      body: {
        userId,
        todoId: input.todoId,
        pomodoroId: input.pomodoroId,
        action: "pause",
      },
    },
  }).response;

  if (statusCode !== 200) return null;
  const json = (await body.json()) as { elapsedSeconds?: number };
  return json.elapsedSeconds ?? null;
}

export async function resetPomodoro(input: {
  pomodoroId: string;
  todoId: string;
}): Promise<boolean> {
  const userId = await getUserId();
  if (!userId) return false;

  const headers = await bearerAuthHeaders();
  const { statusCode } = await put({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      body: {
        userId,
        todoId: input.todoId,
        pomodoroId: input.pomodoroId,
        action: "reset",
      },
    },
  }).response;

  return statusCode === 200;
}

export async function completePomodoro(input: {
  pomodoroId: string;
  todoId: string;
}): Promise<boolean> {
  const userId = await getUserId();
  if (!userId) return false;

  const headers = await bearerAuthHeaders();
  const { statusCode } = await put({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      body: {
        userId,
        todoId: input.todoId,
        pomodoroId: input.pomodoroId,
        action: "complete",
      },
    },
  }).response;

  return statusCode === 200;
}

export async function deletePomodoro(input: {
  pomodoroId: string;
  todoId: string;
}): Promise<boolean> {
  const userId = await getUserId();
  if (!userId) return false;

  const headers = await bearerAuthHeaders();
  const { statusCode } = await del({
    apiName: "CountdownApi",
    path: "/pomodoros",
    options: {
      headers,
      queryParams: {
        userId,
        todoId: input.todoId,
        pomodoroId: input.pomodoroId,
      },
    },
  }).response;

  return statusCode === 200;
}
